Allow use of already marked input files with frr-reload.py

deastoe opened the pull request:
Add a new `--marked` argument to frr-reload.py which indicates that
the new config file (and test running config file, if provided) are
already marked.
    
In this scenario frr-reload.py will directly read the files, and not
via `vtysh -m`. The marking functionality does not always mark the
end of sections correctly and can be dependent on section ordering
in the desired config file.
    
For example, the following snippet:
```    
      router bgp 65000
      end
      !
      vrf foo
      ip route 10.10.0.0/24 blackhole
      exit-vrf
```
is not marked correctly:
```
      line 6: % Unknown command: exit-vrf
    
      router bgp 65000
      !
      vrf foo
      end
      ip route 10.10.0.0/24 blackhole
```

Issue #7326

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/3cd09ce4b24e3fdcf751a4d8fefd1fda/raw/5bafa68c76641a60102ad2c21f2dbd8c74cdf0f6/cr_7327_1602862112.diff | git apply
```

</p>

<p>

```diff
diff --git a/tools/frr-reload.py b/tools/frr-reload.py
index 2e2f3163a..522b1bdd3 100755
--- a/tools/frr-reload.py
+++ b/tools/frr-reload.py
@@ -186,17 +186,17 @@ class Vtysh(object):
 class Context(object):
 
     """
-    A Context object represents a section of frr configuration such as:
-!
-interface swp3
- description swp3 -> r8's swp1
- ipv6 nd suppress-ra
- link-detect
-!
+        A Context object represents a section of frr configuration such as:
+    !
+    interface swp3
+     description swp3 -> r8's swp1
+     ipv6 nd suppress-ra
+     link-detect
+    !
 
-or a single line context object such as this:
+    or a single line context object such as this:
 
-ip forwarding
+    ip forwarding
 
     """
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/950ce5674a6aa8941b3c9c54c9773b09/raw/5bafa68c76641a60102ad2c21f2dbd8c74cdf0f6/cr_7327_1602862172.diff | git apply
```

</p>

<p>

```diff
diff --git a/tools/frr-reload.py b/tools/frr-reload.py
index 2e2f3163a..522b1bdd3 100755
--- a/tools/frr-reload.py
+++ b/tools/frr-reload.py
@@ -186,17 +186,17 @@ class Vtysh(object):
 class Context(object):
 
     """
-    A Context object represents a section of frr configuration such as:
-!
-interface swp3
- description swp3 -> r8's swp1
- ipv6 nd suppress-ra
- link-detect
-!
+        A Context object represents a section of frr configuration such as:
+    !
+    interface swp3
+     description swp3 -> r8's swp1
+     ipv6 nd suppress-ra
+     link-detect
+    !
 
-or a single line context object such as this:
+    or a single line context object such as this:
 
-ip forwarding
+    ip forwarding
 
     """
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

deastoe commented:
ci:rerun

edipascale commented:
The commits are not properly split - the comment introduced by the first commit implies the existence of the marked parameter, which is only added in the second commit. I would just squash everything into a single commit, it's a small enough change.

As for the actual problem the patch is trying to address, personally I have no objection in adding another knob, but shouldn't we try to fix `vtysh -m` instead/as well?

Lastly, please add documentation for this option in the user docs: http://docs.frrouting.org/en/latest/frr-reload.html

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/268611031575834d65bf4ec609ca95dd/raw/5bafa68c76641a60102ad2c21f2dbd8c74cdf0f6/cr_7327_1603359195.diff | git apply
```

</p>

<p>

```diff
diff --git a/tools/frr-reload.py b/tools/frr-reload.py
index 2e2f3163a..522b1bdd3 100755
--- a/tools/frr-reload.py
+++ b/tools/frr-reload.py
@@ -186,17 +186,17 @@ class Vtysh(object):
 class Context(object):
 
     """
-    A Context object represents a section of frr configuration such as:
-!
-interface swp3
- description swp3 -> r8's swp1
- ipv6 nd suppress-ra
- link-detect
-!
+        A Context object represents a section of frr configuration such as:
+    !
+    interface swp3
+     description swp3 -> r8's swp1
+     ipv6 nd suppress-ra
+     link-detect
+    !
 
-or a single line context object such as this:
+    or a single line context object such as this:
 
-ip forwarding
+    ip forwarding
 
     """
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

deastoe commented:
Thanks for the review @edipascale - I've squashed the commits and updated the docs.

I agree that ideally `vtysh -m` would always do the right thing (I partially opened this PR to start the discussion, in addition to providing a workaround). I left some thoughts on the linked ticket - I think there could be some corner cases where the automatic marking cannot always know the right thing to do (ie. we could fix this issue, but that might expose some other issues/corner cases).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/656106ca618f0377b4287f9c0e68d8eb/raw/5bafa68c76641a60102ad2c21f2dbd8c74cdf0f6/cr_7327_1603359767.diff | git apply
```

</p>

<p>

```diff
diff --git a/tools/frr-reload.py b/tools/frr-reload.py
index 2e2f3163a..522b1bdd3 100755
--- a/tools/frr-reload.py
+++ b/tools/frr-reload.py
@@ -186,17 +186,17 @@ class Vtysh(object):
 class Context(object):
 
     """
-    A Context object represents a section of frr configuration such as:
-!
-interface swp3
- description swp3 -> r8's swp1
- ipv6 nd suppress-ra
- link-detect
-!
+        A Context object represents a section of frr configuration such as:
+    !
+    interface swp3
+     description swp3 -> r8's swp1
+     ipv6 nd suppress-ra
+     link-detect
+    !
 
-or a single line context object such as this:
+    or a single line context object such as this:
 
-ip forwarding
+    ip forwarding
 
     """
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

deastoe commented:
Closing this as, following further testing, this does not actually completely address the issue.
